proposals/dht-js.md

```diff
@@ -0,0 +1,106 @@
+# Effective content and service discovery in JS
+
+Authors: @jacobheun
+
+Initial PR: TBD
+
+## Purpose &amp; impact
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives.
+-->
+
+**Current State**
+- JS projects rely on delegate and preload nodes to be able to interact with the live network
+- PL hosted delegate/preload nodes are often overloaded negatively impact performance of projects
+
+#### Assumptions &amp; hypotheses
+_What must be true for this project to matter?_
+<!--(bullet list)-->
+
+TODO
+
+#### User workflow example
+_How would a developer or user use this new capability?_
+<!--(short paragraph)-->
+
+TODO
+
+#### Impact
+TODO
+
+<!--
+Explain why you have chosen this rating
+What awesome potential impact/outcomes/results will we see if we nail this project?
+-->
+
+#### Leverage
+TODO
+
+<!-- Explain the opportunity or leverage point for our subsequent velocity/impact (e.g. by speeding up development, enabling more contributors, etc)
+-->
+
+#### Confidence
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+<!--Explain why this rating-->
+
+
+## Project definition
+#### Brief plan of attack
+
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+The focus of this effort is to build the DHT to function solely in Node.js. Running a DHT in browser is not currently viable.
```

Comment:
Here's a stupid question from someone with little exposure to js ipfs or js ecosystem at large.  What is the benefit to building out functionality in node.js as opposed to relying on a delegate go-ipfs nodes?  Does the value come from having multiple fully functional implementations?  Is there a positive impact on in-browser nodes from delegating DHT functionality to a node.js ipfs node instead of a go-ipfs node?

Reply:
Not a stupid question, and both are viable options with pros/cons and I think there is merit in both. I will expand more here when I have time to flush this out, but I think there is significant work that needs to be done improving remote api access to Go nodes and that should be a separate proposal. There is a lot to be gained there for new developers. The DHT in JS is relatively short term in comparison and immediately improves usability of the JS ecosystem of projects for existing developers (we currently have users leveraging JS and Go in hacky ways to get around some limitations in these systems; flexible IPLD in JS, performant DHT in Go).